### ISSUE TITLE
Research Scaling GPUs

### ISSUE THREAD

**Issue by new5558** [OpenThaiGPT/openthaigpt-pretraining#3]
3 วิธีหลักในการสเกลโมเดล 3D Parallelism:
- Advanced data parallelism เจอว่ามีอะไรที่เหนือว่า DDP ปกติด้วย (FSDP, Sharded DDP, Zero-2, Zero-3, Activation checkpointing, LAMB, 1 bit Adam)
- Pipeline parallelism อันนี้ผล benchmark flops เด็ดมาก (Gpipe, Pipedream)
- Model/Tensor parallelism (Megatron)

เทคนิคอื่นๆ:
- CPU offloading สำหรับเทรนโมเดลใหญ่ใน GPU เครื่องเดียว
- Gradient accumulation ที่มาในฐานะ hyperparamters ที่ส่งผลอย่างมากกับ Compute Flops Efficiency มากๆเมื่อใช่ร่วมกับ 3D Parallelism เวลาเทรนกับ Cluster ขนาดใหญ่

**Comment by new5558** [comment]
volunteer
